### PR TITLE
filter: add language http filter "accept-language"

### PR DESCRIPTION
**Pull request by diceride**
Commit Message:
filter: add a new anguage filter "accept-language"

Additional Description:
The language detection filter (i18n) picks the best match between a client's desired locales and an application's supported locales.

Parses a list of locales from an ``Accept-Language`` header [RFC 2616 Section 14.4](https://tools.ietf.org/html/rfc2616#section-14.4)
and adds a new `x-language` header to the request.

The filter uses [Unicode ICU](https://github.com/unicode-org) for locale parsing.

```txt
    // Multiple types, weighted with the quality value syntax:
    Accept-Language: fr-CH, fr;q=0.9, en;q=0.8, *;q=0.5
```

* The client is from the Romandy region and prefers Swiss French, the variety of French spoken in the French-speaking area of Switzerland
* The client sends an ``Accept-Language`` header
* An ``Accept-Language`` header indicates the natural language and locale that the client prefers
* The application is only configured to match French fr
* The filter sets the value of the ``x-language`` header to fr

The language detection filter outputs statistics in the *http.<stat_prefix>.language.* namespace.

Risk Level:
Low

Testing:
Adds integration tests

Docs Changes:
Adds a new documentation page for the language filter

Release Notes:

Platform Specific Features:

### CODE REVIEW

**Comment by diceride** [comment]
### Example configuration

Minimal filter configuration:

```yaml
  name: envoy.filters.http.language
  typed_config:
    "@type": type.googleapis.com/envoy.extensions.filters.http.language.v1.Language
    default_language: en
    supported_languages: [en, en-uk, de, dk, es, fr, zh, zh-tw]
```

The above configuration can be understood as follows:

* Try to pick the client's desired locale from an ``Accept-Language`` header
* At any point the filter uses [Unicode ICU](https://github.com/unicode-org) for locale parsing
* If the client's desired locale can not be picked, for example because the client provided an invalid value, the `Language.default_language` option will be used as a fallback

**Comment by diceride** [comment]
Example configuration

Minimal filter configuration:

```yaml
  name: envoy.filters.http.language
  typed_config:
    "@type": type.googleapis.com/envoy.extensions.filters.http.language.v1.Language
    default_language: en
    supported_languages: [en, en-uk, de, dk, es, fr, zh, zh-tw]
```

The above configuration can be understood as follows:

* Try to pick the client's desired locale from an ``Accept-Language`` header
* At any point the filter uses [Unicode ICU](https://github.com/unicode-org) for locale parsing
* If the client's desired locale can not be picked, for example because the client provided an invalid value, the `Language.default_language` option will be used as a fallback

File: api/envoy/extensions/filters/http/language/v3/language.proto
Comment (adisuissa) [inline api/envoy/extensions/filters/http/language/v3/language.proto:31]: Must there be at least one supported language? if so, consider adding `min_items` constraint.

File: api/envoy/extensions/filters/http/language/v3/language.proto
Comment (adisuissa) [inline api/envoy/extensions/filters/http/language/v3/language.proto:24]: General comment about all the fields that expect a well-known string, please validate their correctness when ingesting the config, and reject that config if it contains a wrong field.

**Review by adisuissa — COMMENTED** [review]
Thanks!
I've left a couple of API comments.

IIRC the ICU library was previously used in Envoy, but later removed.
Not sure what the details are, but maybe @wrowe as a dependency-reviewer can shed some light.

File: api/envoy/extensions/filters/http/language/v3/language.proto
Comment (diceride) [reply api/envoy/extensions/filters/http/language/v3/language.proto]: Updated the validator

```proto
repeated string supported_languages = 2 [(validate.rules).repeated = {
  min_items: 1
  unique: true
  items {string {min_len: 2}}
}];
```

**Comment by adisuissa** [comment]
@htuch before I take another pass on this, can you remind me why icu was removed?

**Comment by diceride** [comment]
I'm not familiar with Windows. The msvc-cl pipeline fails to compile libicu with the following error

```txt
checking whether the C compiler works... no
...
configure: error: C compiler cannot create executables
```

File: api/envoy/extensions/filters/http/language/v3/language.proto
Comment (diceride) [reply api/envoy/extensions/filters/http/language/v3/language.proto:23]: Added additional config validators to the protocol buffer definition and to LanguageFilterFactory::createFilterFactoryFromProtoTyped as suggested.

Comment (diceride) [reply api/envoy/extensions/filters/http/language/v3/language.proto:23]: Added additional validators to the protocol buffer definition and to LanguageFilterFactory::createFilterFactoryFromProtoTyped as suggested.

Comment (diceride) [reply api/envoy/extensions/filters/http/language/v3/language.proto:23]: Added additional validators to the protocol buffer definition and to LanguageFilterFactory::createFilterFactoryFromProtoTyped.

**Comment by htuch** [comment]
ICU is pretty scary from a security perspective, see the numerous CVEs https://www.cvedetails.com/vulnerability-list/vendor_id-17477/Icu-project.html. @yanavlasov also has some experience with this, there is substantial history in the non-public https://github.com/envoyproxy/envoy-setec/issues/196.

Not super excited to bring this back into the Envoy repository TBH, but willing to be persuaded.

**Comment by diceride** [comment]
@htuch The extension only uses a very limited subset of ICU, in particular icu::Locale and the icu::LocaleMatcher. ICU 67.1 (launched in April 2020) comes with an improved icu::LocaleMatcher API and implementation. The generated ICU data file is minimal.

**Comment by htuch** [comment]
@diceride that reminds me that we've had some outages with ICU artifacts that were supposed to be linked in but weren't present on the system. I'm going to defer to @yanavlasov here as he is ICU expert and also it would be good to hear from other @envoyproxy/dependency-shepherds like @moderation @wrowe. If there is consensus that a limited use of ICU for this application is OK that's fine by me, but I'm a little concerned we're undoing a deliberative historical choice on deps here.

**Comment by wrowe** [comment]
> If there is consensus that a limited use of ICU for this application is OK that's fine by me, but I'm a little concerned we're undoing a deliberative historical choice on deps here.

If we are talking about parsing locale data, or simply transforming between charsets, ICU is fine and more portable on Windows than the GPL iconv. If we bring this in we might want to cautiously limit what components we configure for, much like we've done for curl. The windows build could be adjusted, ping me, but as long as we follow the RFC our own implementation of locale code matching and basic validation would be trivial.

ICU is maintained and as close to a reference implementation as we can get.

**Comment by wrowe** [comment]
> The windows build could be adjusted

Note that our convention has been to use CMake (foreign_cc) builds 
in the absence of bazel-build for these dependencies. Autoconf has
been strongly avoided. I'll have a look to see if the current ICU release
can be realistically built on win32, or if this filter simply won't be built
on windows at all.

File: source/extensions/extensions_metadata.yaml
Comment (diceride) [reply source/extensions/extensions_metadata.yaml:229]: I changed the security posture to `requires_trusted_downstream_and_upstream`.

Comment (diceride) [reply source/extensions/extensions_metadata.yaml:229]: Updated the security posture to `requires_trusted_downstream_and_upstream`.

File: bazel/repositories.bzl
Comment (diceride) [reply bazel/repositories.bzl:370]: Limited the visibility of the icu library to `//source/extensions/filters/http/language:language_filter_lib` in the bazel/foreign_cc/BUILD file.

Comment (diceride) [reply bazel/repositories.bzl:370]: Limited the visibility of the icu library to `//source/extensions/filters/http/language:__pkg__` in the bazel/foreign_cc/BUILD file and removed the bind target alias to the //external package.

File: source/extensions/extensions_metadata.yaml
Comment (diceride) [reply source/extensions/extensions_metadata.yaml:229]: Added additional character validation to ensure the only permitted characters for values are ASCII letters, hyphen (‘-‘), underscore (‘_’), at-sign (‘@’), equals sign (‘=’), and semicolon (‘;’).

File: source/extensions/filters/http/language/language_filter.cc
Comment (diceride) [reply source/extensions/filters/http/language/language_filter.cc]: Refactored some repeated code into a new isLanguageTagValid method.

File: api/envoy/extensions/filters/http/language/v3/language.proto
Comment (diceride) [reply api/envoy/extensions/filters/http/language/v3/language.proto:34]: The query_param_name and cookie_name are not mutually exclusive, but optional.

Updated the documentation to clarify this point

```txt
...
If enabled, the URL query parameter will be used firrst and the cookie value will be used second to match the desired locale of a client, before
continuing to parse a list of locales from an ``Accept-Language`` header by default.
...
```

Comment (diceride) [reply api/envoy/extensions/filters/http/language/v3/language.proto:34]: The query_param_name and cookie_name are not mutually exclusive, but optional.

Updated the documentation to clarify this point

```txt
...
If enabled, the URL query parameter will be used first and the cookie value will be used second to match the desired locale of a client, before
continuing to parse a list of locales from an ``Accept-Language`` header by default.
...
```

**Comment by diceride** [comment]
This filter now only matches the desired locale of a client using the Accept-Language header and nothing more. The support to match the desired locale of a client using a query parameter value or a cookie value was removed.

This simplifies the API and improves security, as only the [getBestMatchForListString()](https://unicode-org.github.io/icu-docs/apidoc/dev/icu4c/classicu_1_1LocaleMatcher.html#a6d3388c67d00686a78e962c765ca7a5e) method of the icu [LocaleMatcher](https://unicode-org.github.io/icu-docs/apidoc/dev/icu4c/classicu_1_1LocaleMatcher.html) class is used.

File: api/envoy/extensions/filters/http/language/v3/language.proto
Comment (diceride) [reply api/envoy/extensions/filters/http/language/v3/language.proto]: Removed the cookie and query parameter options. See comment below.

**Comment by diceride** [comment]
This filter now only matches the desired locale of a client using the Accept-Language header and nothing more. The support to match the desired locale of a client using a query parameter value or a cookie value was removed.

This simplifies the API and improves security, as only the [getBestMatchForListString()](https://unicode-org.github.io/icu-docs/apidoc/dev/icu4c/classicu_1_1LocaleMatcher.html#a6d3388c67d00686a78e962c765ca7a5e) method of the icu [LocaleMatcher](https://unicode-org.github.io/icu-docs/apidoc/dev/icu4c/classicu_1_1LocaleMatcher.html) class is used.

**Comment by diceride** [comment]
The support to match the desired locale of a client using a query parameter value or a cookie value was removed.

This filter now only matches the desired locale of a client using the Accept-Language header. It simplifies the API and improves security, as only the [getBestMatchForListString()](https://unicode-org.github.io/icu-docs/apidoc/dev/icu4c/classicu_1_1LocaleMatcher.html#a6d3388c67d00686a78e962c765ca7a5e) method of the icu [LocaleMatcher](https://unicode-org.github.io/icu-docs/apidoc/dev/icu4c/classicu_1_1LocaleMatcher.html) class is used.

File: CODEOWNERS
Comment (yanavlasov) [inline CODEOWNERS:95]: Have you coordinated with @mattklein123 regarding sponsorship? If not, then you either need to have maintainer sponsor or move the filter code into the `contrib` directory, per the contribution policy.

File: source/extensions/filters/http/language/language_filter.cc
Comment (yanavlasov) [inline source/extensions/filters/http/language/language_filter.cc:53]: You can not throw from the filter constructor as we avoid using exceptions on the dataplane threads. You will need to create this matcher eiother as part of the factory or configuration, so that an error and exception can be handled during config processing.

**Review by yanavlasov — CHANGES_REQUESTED** [review]
/wait

File: source/extensions/filters/http/language/language_filter.cc
Comment (diceride) [reply source/extensions/filters/http/language/language_filter.cc]: I did not know about that. Moved the code to the factory implementation.

File: CODEOWNERS
Comment (diceride) [reply CODEOWNERS]: I did not know how to proceed here as at least two contributors are required. For now, I added you @yanavlasov to the contrib section of the CODEOWNERS file. Please let me know if this is all right with you. Otherwise I don't know how to proceed here.

```txt
/contrib/language/ @diceride @yanavlasov
```

**Comment by diceride** [comment]
* Moved the extension to the contrib directory and registered it as a contrib extension
* Renamed the API version from v3 to v3alpha
* Disabled Windows support
* Updated the documentation and added a warning that Windows support is disabled
* Updated the documentation and added a warning that this filter is experimental

File: CODEOWNERS
Comment (diceride) [reply CODEOWNERS]: I did not know whom to add as at least two contributors are required. For now, I added you @yanavlasov to the contrib section of the CODEOWNERS file. Please let me know if this is all right with you. Otherwise I don't know how to proceed here.

```txt
/contrib/language/ @diceride @yanavlasov
```

**Comment by diceride** [comment]
* Moved the extension to the contrib directory and registered it as a contrib extension
* Renamed the API version from v3 to v3alpha
* Disabled Windows support
* Updated the documentation and added a warning that Windows support is disabled
* Updated the documentation and added a warning that this filter is experimental

**Comment by diceride** [comment]
Added a new method validateHeaderValue to determine whether the Accept-Language header value is compliant with the specification before parsing the header value using the [getBestMatchForListString()](https://unicode-org.github.io/icu-docs/apidoc/dev/icu4c/classicu_1_1LocaleMatcher.html#a6d3388c67d00686a78e962c765ca7a5e) method of the icu [LocaleMatcher](https://unicode-org.github.io/icu-docs/apidoc/dev/icu4c/classicu_1_1LocaleMatcher.html) class.

File: contrib/language/filters/http/source/language_filter.cc
Comment (adisuissa) [inline contrib/language/filters/http/source/language_filter.cc:56]: nit:
```suggestion
    supported_languages_.push_back(locale);
```

**Review by adisuissa — CHANGES_REQUESTED** [review]
Thanks for working on this!
Left a few comments.
/wait

File: CODEOWNERS
Comment (yanavlasov) [inline CODEOWNERS:222]: `contrib` extensions do not need maintainer sponsor. Please just leave your username.

File: contrib/language/filters/http/source/language_filter.cc
Comment (yanavlasov) [inline contrib/language/filters/http/source/language_filter.cc:82]: Note, that this will cause abnormal process termination, since this exception will not be handled on the worker thread. You should move matcher creation to the filter config object. Note that matcher object should be readonly and possible to use concurrently.

**Review by yanavlasov — CHANGES_REQUESTED** [review]
LGTM module remaining comments.

Note that I will be unable to provide ownership to this extension due to lack of time and for a contrib extension you can just put your name as the sole owner.

/wait

File: contrib/language/filters/http/source/config.cc
Comment (diceride) [reply contrib/language/filters/http/source/config.cc:19]: Unfortunately, we have to keep this shared to access the stats method. 

Example

```
config_->stats().default_language_.inc();
```

File: api/contrib/envoy/extensions/filters/http/language/v3alpha/language.proto
Comment (diceride) [reply api/contrib/envoy/extensions/filters/http/language/v3alpha/language.proto:20]: I went deep into the ICU code, as I was not sure myself (it's nowhere documented). The default_language can be exclusive and does not need to be part of the supported languages list. 

But I agree with you that this is an unusual and confusing configuration. Why would a user specify the default_language but exclude it from the supported_language list?

I changed the implementation and the default_language is now added to the supported_language list. I also added some more comments to the protocol buffer API.

**Comment by daixiang0** [comment]
@diceride please fix CI.

File: CODEOWNERS
Comment (diceride) [reply CODEOWNERS]: The CI is failing due to the following error

```
ERROR: Contrib extensions require at least 2 owners in CODEOWNERS:
```

It appears that it is not possible to only specify a single codeowner for a contrib extension. Do you know a way to disable this warning?

**Comment by diceride** [comment]
The CI is failing due to the following error

```
ERROR: Contrib extensions require at least 2 owners in CODEOWNERS:
```

File: contrib/language/filters/http/source/language_filter.cc
Comment (diceride) [reply contrib/language/filters/http/source/language_filter.cc:82]: I'm looking into this.

**Comment by diceride** [comment]
* Moved the initialization of the default locale and the locale matcher to the filter config object
* Made the default locale and the locale matcher immutable
* Added a new filter config test
* Updated the CODEOWNERS file to bypass the missing sponsor warning (self-sponsoring)

**Comment by diceride** [comment]
* Moved the initialization of the default locale and the locale matcher to the filter config object
* Made the default locale and the locale matcher immutable
* Updated the CODEOWNERS file to bypass the missing sponsor warning (self-sponsoring)

**Comment by diceride** [comment]
My computer can't handle the workload to run these tests so I'm outsourcing it to the CI for the moment.

**Comment by diceride** [comment]
My computer can't handle the workload to compile these tests so I'm outsourcing it to the CI for the moment.

File: contrib/language/filters/http/source/config.cc
Comment (adisuissa) [inline contrib/language/filters/http/source/config.cc:25]: should this be a set?

File: api/contrib/envoy/extensions/filters/http/language/v3alpha/language.proto
Comment (adisuissa) [inline api/contrib/envoy/extensions/filters/http/language/v3alpha/language.proto:26]: nit: 
```suggestion
  // The set of supported languages. There is no order priority.
```

File: contrib/language/filters/http/source/config.cc
Comment (diceride) [reply contrib/language/filters/http/source/config.cc]: Replaced the implementation with a std::set of icu::Locale and added a custom comparator.

https://unicode-org.github.io/icu-docs/apidoc/dev/icu4c/classicu_1_1Locale.html#a9423f4eb4b671eec3d98c26ec923351e

File: contrib/language/filters/http/source/config.cc
Comment (adisuissa) [inline contrib/language/filters/http/source/config.cc:29]: nit: I think std::unordered_set makes more sense here.

**Review by adisuissa — COMMENTED** [review]
Thanks!
/lgtm api
Left a small nit on the code.
Please avoid force-pushing as it removes comments history (see https://github.com/envoyproxy/envoy/blob/main/CONTRIBUTING.md#submitting-a-pr)

File: contrib/language/filters/http/source/config.cc
Comment (diceride) [reply contrib/language/filters/http/source/config.cc]: Replaced std::set with absl::flat_hash_set.

**Review by adisuissa — DISMISSED** [review]
LGTM thanks!

@yavlasov any other comments?

**Comment by diceride** [comment]
There was a typo in the handle.

@yanavlasov

**Comment by diceride** [comment]
@yanavlasov @wrowe

**Comment by diceride** [comment]
@htuch there is a remaining comment by @adisuissa for the bazel/repository_locations.bzl file. Can you take a look whether the use_category property should be changed?

**Comment by diceride** [comment]
@htuch there is a remaining comment by @adisuissa for the bazel/repository_locations.bzl file. Can you take a look whether the use_category property should be changed?

File: bazel/repository_locations.bzl
Comment (htuch) [reply bazel/repository_locations.bzl:285]: Yeah, I think this would be desirable. It can be a separate PR as a followup, but we should ideally be able to better track which deps are used by core, extensions and contrib, since there are different security implications in the threat model. CC @envoyproxy/dependency-shepherds

**Comment by htuch** [comment]
@diceride I left a comment - we don't need to block merging this PR on that, but we should follow up ideally.

**Comment by diceride** [comment]
@htuch I will follow up on this in a separate PR as suggested by you in https://github.com/envoyproxy/envoy/pull/20143#discussion_r839147761.

@yanavlasov Ready to be reviewed again.

File: contrib/language/filters/http/source/config.cc
Comment (yanavlasov) [reply contrib/language/filters/http/source/config.cc:61]: nit: Envoy style prefers `std::make_shared<LanguageFilter>(config)` instead of two step using operator new.

**Review by yanavlasov — APPROVED** [review]
LGTM modulo small nit.

/wait

File: contrib/language/filters/http/source/config.cc
Comment (diceride) [reply contrib/language/filters/http/source/config.cc]: Done

**Comment by diceride** [comment]
Restarted the CI as the Azure build pipeline ran out of disk space the last time.

**Comment by diceride** [comment]
One of the Azure build pipeline steps ran out of disk space. Force-pushed to restart the CI.

**Comment by diceride** [comment]
One of the Azure build pipeline steps ran out of disk space. Force-pushed to restart the CI.

**Comment by diceride** [comment]
One of the Azure build pipeline steps ran out of disk space multiple times now. Rebased and force-pushed to restart the CI.

**Comment by diceride** [comment]
https://github.com/envoyproxy/envoy/issues/20619

**Comment by wrowe** [comment]
> Rebased and force-pushed to restart the CI.

Please, do not rebase. You can simply push an empty commit (--allow-empty) to shake the CI.

**Comment by diceride** [comment]
One of the Azure build pipeline steps is still running out of disk space.

**Comment by diceride** [comment]
Added a new commit to prevent one of the Azure build pipeline steps to run out of disk space.

@wrowe @yanavlasov

**Comment by diceride** [comment]
Added a new commit to prevent one of the Azure build pipeline steps from running out of disk space.

@wrowe @yanavlasov

**Review by yanavlasov — APPROVED** [review]

**Comment by yanavlasov** [comment]
This is good to go. Please resolve conflicts and then I will merge.

**Comment by diceride** [comment]
@wrowe I know that. I squashed some previous empty commits that had missing or wrong DCO signatures, without rewriting the tree to the point of the changes.

**Comment by diceride** [comment]
@wrowe I know that. I squashed some previous empty commits that had missing or wrong signatures, while only rewriting the tree to the point of changes as the DCO check was failing.

**Comment by diceride** [comment]
@wrowe I know that. I squashed some previous empty commits that had missing or wrong signatures, while only rewriting the tree to the point of changes as the DCO check was failing. I did not know that I can issue commands to repokitteh to retry single build steps, instead of using empty commits.

@yanavlasov I resolved the merge conflicts, can you check again? Thank you.

**Comment by diceride** [comment]
@wrowe I know that. I squashed some previous empty commits that had missing or wrong signatures, while only rewriting the tree to the point of changes as the DCO check was failing. I did not know that I can issue commands to repokitteh to retry single build steps, instead of using empty commits.

@yanavlasov I resolved the merge conflicts.

**Comment by phlax** [comment]
@diceride can you merge main - it seems like there are some changes in this PR that are already present on main

**Comment by yanavlasov** [comment]
I'll merge once the change in .azure-pipelines/pipelines.yml is resolved

/wait

**Comment by diceride** [comment]
Dropped the pipeline commits and merged with main.

**Comment by sc0ttbeardsley** [comment]
Great work! We had a similar need and didn't get around to adding this functionality yet. Thanks!

**Pull request merged**
